plugin/src/main/java/org/opensearch/ml/action/models/UpdateModelTransportAction.java

```diff
@@ -0,0 +1,326 @@
+                                    } else {
+                                        actionListener
+                                            .onFailure(
+                                                new MLValidationException(
+                                                    "User doesn't have privilege to perform this operation on this model, model ID "
+                                                        + modelId
+                                                )
+                                            );
+                                    }
+                                }, exception -> {
+                                    log.error("Permission denied: Unable to update the model with ID {}. Details: {}", modelId, exception);
+                                    actionListener.onFailure(exception);
+                                }));
+                        } else {
+                            actionListener
+                                .onFailure(
+                                    new MLValidationException(
+                                        "User doesn't have privilege to perform this operation on this function category: "
+                                            + mlModel.getAlgorithm().toString()
+                                    )
+                                );
+                        }
+                    } catch (Exception e) {
+                        log.error("Failed to update ML model for model ID {}. Details {}:", modelId, e);
+                        actionListener.onFailure(e);
+                    }
+                } else {
+                    actionListener
+                        .onFailure(new IllegalArgumentException("Failed to find model to update with the provided model id: " + modelId));
+                }
+            }, e -> actionListener.onFailure(new MLResourceNotFoundException("Fail to find model"))));
+        } catch (Exception e) {
+            log.error("Failed to update ML model for " + modelId, e);
+            actionListener.onFailure(e);
+        }
+    }
+
+    private void updateUndeployedModel(
+        String modelId,
+        MLUpdateModelInput updateModelInput,
+        MLModel mlModel,
+        User user,
+        UpdateRequest updateRequest,
+        ActionListener<UpdateResponse> actionListener,
+        ThreadContext.StoredContext context
+    ) {
+        SearchRequest searchRequest = new SearchRequest(ML_MODEL_INDEX);
+        SearchSourceBuilder sourceBuilder = new SearchSourceBuilder();
+        BoolQueryBuilder boolQueryBuilder = QueryBuilders.boolQuery();
+        boolQueryBuilder.must(QueryBuilders.matchQuery(MLModel.MODEL_ID_FIELD, modelId));
+        boolQueryBuilder.must(QueryBuilders.idsQuery().addIds(mlModelManager.getAllModelIds()));
+        sourceBuilder.query(boolQueryBuilder);
+        searchRequest.source(sourceBuilder);
+
+        client.search(searchRequest, ActionListener.wrap(searchResponse -> {
```

Comment:
might need another stashContext/restoreContext here depending how you resolve https://github.com/opensearch-project/ml-commons/pull/1350/files#r1344812309

Reply:
oh I see that gets passed through the entire chain and restored at the very end, nvm

Reply:
tho I think ActionListener.onFailure doesn't throw its exception so the unhappy cases may still drop the thread?